std/experimental/bigo.d

```diff
@@ -0,0 +1,399 @@
+module std.experimental.bigo;
+
+struct BigO
+{
+    import std.typecons : Tuple;
+    import std.algorithm : cartesianProduct, map, sum;
+    import std.array : array, empty, front, popFront;
+    import std.range : chain;
+
+    alias Atom = Tuple!(int, "id", double, "exp", double, "logExp");
+    private Atom[][] terms;
+
+    string toString() const @safe
+    {
+        if (terms.empty) return "O(1)";
+        string result = "O(";
+        foreach (ref e; terms)
+        {
+            if (result.length > 2) result ~= " + ";
+            foreach (ref f; e)
+            {
+                import std.conv : to;
+                if (f.exp != 0)
+                {
+                    result ~= "n";
+                    if (f.id != 0) result ~= f.id.to!string;
+                    if (f.exp != 1)
+                    {
+                        result ~= "^^";
+                        result ~= f.exp.to!string;
+                    }
+                }
+                if (f.logExp != 0)
+                {
+                    if (f.exp != 0)
+                    {
+                        result ~= " * ";
+                    }
+                    result ~= "log(n";
+                    if (f.id) result ~= f.id.to!string;
```

Comment:
It's a micro optimization to be sure, but if you use appender in this function, then instead of

```
result ~= f.id.to!string;
```

which allocates a temporary array then throws it away, you can do

```
result.put(f.id.toChars);
```

which will only use appender's internal buffer